### ISSUE TITLE
Pass automated tests for zig stdlib

### ISSUE THREAD

**Issue by codefromthecrypt** [tetratelabs/wazero#1029]
**Is your feature request related to a problem? Please describe.**
One of our goals of 1.0 is to reduce the amount of glitches using syscall ABIs, notably WASI. While there is an emerging [wasi-testsuite](https://github.com/WebAssembly/wasi-testsuite), and we have a number of ad-hoc tests, we should also pass zig tests as an integration suite.

**Describe the solution you'd like**

#1029 runs automated tests, but we don't yet pass them

from zig clone
```bash
$ zig test --test-cmd wazero --test-cmd run --test-cmd -mount=.:/ --test-cmd-bin -target wasm32-wasi --zig-lib-dir ./lib ./lib/std/std.zig
```

The actual mounts may need to vary once above actually passes. e.g. mounting tmp for writes

**Describe alternatives you've considered**

We currently do this manually and backfill notable gaps as unit tests, but the logic zig uses changes from time to time.

**Additional context**

As of Jan 13, this doesn't proceed as we've not fully implemented wasi syscalls needed, nor have the correct flags to proceed. e.g. /tmp
```
$ zig test --test-cmd wazero --test-cmd run --test-cmd -hostlogging=filesystem --test-cmd -mount=.:/ --test-cmd-bin -target wasm32-wasi --zig-lib-dir ./lib ./lib/std/std.zig
unexpected errno: 21
Unable to dump stack trace: not implemented for Wasm
unexpected errno: 21
Unable to dump stack trace: not implemented for Wasm
expected error.EnvironmentVariableNotFound, found error.OutOfMemory
==> wasi_snapshot_preview1.path_create_directory(fd=3,path=zig-cache)
<== errno=EEXIST
==> wasi_snapshot_preview1.path_open(fd=3,dirflags=SYMLINK_FOLLOW,path=zig-cache,oflags=DIRECTORY,fs_rights_base=FDSTAT_SET_FLAGS|PATH_CREATE_DIRECTORY|PATH_CREATE_FILE|PATH_LINK_SOURCE|PATH_LINK_TARGET|PATH_OPEN|FD_READDIR|PATH_READLINK|PATH_RENAME_SOURCE|PATH_RENAME_TARGET|PATH_FILESTAT_GET|PATH_FILESTAT_SET_SIZE|PATH_FILESTAT_SET_TIMES|FD_FILESTAT_GET|FD_FILESTAT_SET_TIMES|PATH_SYMLINK|PATH_REMOVE_DIRECTORY|PATH_UNLINK_FILE,fs_rights_inheriting=FD_DATASYNC|FD_READ|FD_SEEK|FDSTAT_SET_FLAGS|FD_SYNC|FD_TELL|FD_WRITE|FD_ADVISE|FD_ALLOCATE|PATH_CREATE_DIRECTORY|PATH_CREATE_FILE|PATH_LINK_SOURCE|PATH_LINK_TARGET|PATH_OPEN|FD_READDIR|PATH_READLINK|PATH_RENAME_SOURCE|PATH_RENAME_TARGET|PATH_FILESTAT_GET|PATH_FILESTAT_SET_SIZE|PATH_FILESTAT_SET_TIMES|FD_FILESTAT_GET|FD_FILESTAT_SET_SIZE|FD_FILESTAT_SET_TIMES|PATH_SYMLINK|PATH_REMOVE_DIRECTORY|PATH_UNLINK_FILE|POLL_FD_READWRITE,fdflags=)
<== (opened_fd=4,errno=ESUCCESS)
==> wasi_snapshot_preview1.path_create_directory(fd=4,path=tmp)
<== errno=EEXIST
--snip--
```

**Comment by mathetake** [comment]
Changed the title to "pass" as the automated job has been added in https://github.com/tetratelabs/wazero/pull/1031

**Comment by codefromthecrypt** [comment]
it appears that the failing tests are for symbolic and hard links. Note: old_fd and new_fd here are the directories that enclosed the files, not the files themselves

```
--> wasi_snapshot_preview1.path_symlink(old_path=file.txt,fd=9,new_path=symlink1)
293
<-- errno=ENOSYS
--> wasi_snapshot_preview1.path_symlink(old_path=file.txt,fd=8,new_path=link)
213
<-- errno=ENOSYS
--> wasi_snapshot_preview1.path_link(old_fd=3,old_flags=,old_path=example.txt,new_fd=6,new_path=new.txt)
143
<-- errno=ENOSYS
--> wasi_snapshot_preview1.path_link(old_fd=3,old_flags=,old_path=example.txt,new_fd=3,new_path=new.txt)
115
<-- errno=ENOSYS
--> wasi_snapshot_preview1.path_symlink(old_path=file.txt,fd=3,new_path=symlinked)
101
<-- errno=ENOSYS
```

**Comment by codefromthecrypt** [comment]
after https://github.com/ziglang/zig/pull/14372 more pass

**Comment by codefromthecrypt** [comment]
next step is to figure out why this is going back to cookie zero (which is invalid)
```bash
$ /Users/adrian/Downloads/zig-macos-x86_64-0.11.0-dev.1269+c2d37224c/zig test --test-cmd wazero --test-cmd run --test-cmd -hostlogging=filesystem --test-cmd -mount=.:/ --test-cmd-bin -target wasm32-wasi --zig-lib-dir ./lib ./lib/std/std.zig
---snip--
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=0)
<== (bufused=65,errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=2)
<== (bufused=0,errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=0)
<== (bufused=,errno=EINVAL)
reached unreachable codeerror instantiating wasm binary: module[] function[_start] failed: wasm error: unreachable
wasm stack trace:
	.os.abort()
		0x1f5a: /Users/adrian/oss/zig/lib/std/os.zig:553:9
	.builtin.default_panic(i32,i32,i32,i32)
		0xa59: /Users/adrian/oss/zig/lib/std/builtin.zig:863:25
	.fs.IterableDir.Iterator__struct_86503.nextWasi(i32,i32)
		0x1163ff0: /Users/adrian/oss/zig/lib/std/fs.zig:820:39
	.fs.IterableDir.Iterator__struct_86503.next(i32,i32)
		0x1163535: /Users/adrian/oss/zig/lib/std/fs.zig:792:37
	.fs.test.test.Dir.Iterator twice() i32
		0x1166203: /Users/adrian/oss/zig/lib/std/fs/test.zig:241:29
	.test_runner.main2() i32
		0x262: /Users/adrian/oss/zig/lib/test_runner.zig:126:21
	.test_runner.main()
		0x35: /Users/adrian/oss/zig/lib/test_runner.zig:17:21
	._start()
		0x1d: /Users/adrian/oss/zig/lib/std/start.zig:606:22 (inlined)
		      /Users/adrian/oss/zig/lib/std/start.zig:240:42
```

My guess is the problem is around here https://github.com/ziglang/zig/blob/master/lib/std/fs.zig#L832

This should be checking to see if buf_len (input param) is equal to buf_used (output param). These should be the same while there are more entries to read. My guess is what's happening is this is looping on :start_over regardless of this. It should break when buf_used returns less than buf_len.

As you'll notice in the log, it keeps going, as the first read was a full read (bufused = 65 < buf_len = 1024)

```
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=0)
<== (bufused=65,errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=2)
<== (bufused=0,errno=ESUCCESS)
==> wasi_snapshot_preview1.fd_readdir(fd=9,buf=1046228,buf_len=1024,cookie=0)
<== (bufused=,errno=EINVAL)
```

so basically the iteration should have stopped at the first fd_readdir, but it kept going

**Comment by mathetake** [comment]
I will bring this to the finish line

**Comment by mathetake** [comment]
found a fix^ and only 8 tests left 
```
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.symlink with relative paths (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.link with relative paths (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.linkat with different directories (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.readlinkat (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.Dir.readLink (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.. and .. in fs.Dir functions (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.setEndPos (Unexpected)
unexpected errno: 52
Unable to dump stack trace: not implemented for Wasm
FAIL: test.updateTimes (Unexpected)
2263 passed; 142 skipped; 8 failed.
```

**Comment by mathetake** [comment]
4 tests

**Comment by mathetake** [comment]
1 test!

**Comment by mathetake** [comment]
omg passed all

**Comment by mathetake** [comment]
_only on darwin_
